ml/Database.cc

```diff
@@ -38,6 +38,77 @@ const char *ml::Database::SQL_SELECT_ANOMALY_EVENTS =
     "   after >= ?4 AND"
     "   before <= ?5;";
 
+const char *ml::Database::SQL_CREATE_ANOMALY_RATE_INFO_TABLE =
+    "CREATE TABLE IF NOT EXISTS anomaly_rate_info( "
+    "     host_id text NOT NULL, "
+    "     after int NOT NULL, "
+    "     before int NOT NULL, "
+    "     anomaly_rates text "
+    ");";
+
+const char *ml::Database::SQL_INSERT_BULK_ANOMALY_RATE_INFO =
+    "INSERT INTO anomaly_rate_info( "
+    "     host_id, after, before, anomaly_rates) "
+    " VALUES (?1, ?2, ?3, ?4);";
+   
+const char *ml::Database::SQL_SELECT_ANOMALY_RATE_INFO =
+    "SELECT  main.dimension_id, COALESCE("
+    " ((COALESCE(pre.avg,0) * (SELECT before-?2 FROM anomaly_rate_info WHERE after < ?2 ORDER BY after DESC LIMIT 1)) + "
+    " (COALESCE(main.avg,0) * ((SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) - "
+    "              (SELECT after FROM anomaly_rate_info WHERE after >= ?2 ORDER BY after ASC LIMIT 1))) + "
+    " (COALESCE(post.avg,0) * (SELECT ?3-after FROM anomaly_rate_info WHERE before > ?3 OR "
+    "                ((SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) "
+    "                 AND NOT EXISTS (SELECT before FROM anomaly_rate_info WHERE before > ?3 ORDER BY before ASC LIMIT 1)) "
+    "              ORDER BY before ASC LIMIT 1))) / ( "
+    "  (SELECT before-?2 FROM anomaly_rate_info WHERE after < ?2 ORDER BY after DESC LIMIT 1) + "
+    "  (SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) - "
+    "  (SELECT after FROM anomaly_rate_info WHERE after >= ?2 ORDER BY after ASC LIMIT 1) + "
+    "  (SELECT ?3-after FROM anomaly_rate_info WHERE before > ?3 OR "
+    "                 ((SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) "
+    "                  AND NOT EXISTS (SELECT before FROM anomaly_rate_info WHERE before > ?3 ORDER BY before ASC LIMIT 1)) "
+    "               ORDER BY before ASC LIMIT 1) "
+    " ),0.0) percentage FROM "
+    "(SELECT dimension_id, AVG(anomaly_percentage) avg FROM "
```

Comment:
Hi Siamak!

There's also something we might need to check: The case when there is no main part.

Assume our simple example:

|after|before|a_dimension|
|------|---------|---------------|
|10|20|0|
|20|30|5|
|30|40|0|
|40|50|0|

If we request 25 -> 35, then we don't have any main. We do have pre, and post, but since the join is based on the main, we don't get any results to compute.

I'm still investigating it, for a possible solution, but wanted to bring it up for you to also have a look.

Thank you!